framework/src/base/NonlinearSystemBase.C

```diff
@@ -278,6 +278,9 @@ NonlinearSystemBase::addKernel(const std::string & kernel_name, const std::strin
       _time_kernels.addObject(kernel, tid);
     else
       _non_time_kernels.addObject(kernel, tid);
+
+    // store kernels to warehouse, if necessary
```

Comment:
`addExtraWarehouse` is doing nothing in `NonlinearSystemBase`, and its behavior will be implemented in the subclass. 

Adding more methods to warehouse may be  a good idea.